File: python/purehtml/configs/types/primitive_value_config.py
from typing import Any
from typing import List
from typing import Optional

from purehtml.configs.selector_match_params import GetSelectorMatchesParams
from purehtml.configs.types.config_with_selector import ConfigWithSelector
from purehtml.transformers.transform_params import TransformParams
from purehtml.transformers.transformer import Transformer


class PrimitiveValueConfig(ConfigWithSelector):
    def __init__(self, selector: Optional[str], transform: Optional[List[Transformer]] = None):
        """
        Initializes the PrimitiveValueConfig with a selector and optional transformers.
        :param selector: The CSS selector used to find the target element.
        :param transform: A list of Transformer objects (optional).
        """
        super().__init__(selector)
        self.transform = transform if transform else []

    def extract(self, params) -> Any:
        """
        Extracts a value from the parameters, applying any transformations if available.
        :param params: The ExtractParams object containing context for extraction.
        :return: The extracted and transformed value as a string.
        """
        val = None
        # Creating GetSelectorMatchesParams for selector matching
        selector_params = GetSelectorMatchesParams(
            already_matched=params.get_element_already_matched() or False,
            include_root=False,
        )

        selected_element = self.get_first_match(params.node(), selector_params, params.document())
        if selected_element:
            val = selected_element.text()

        if not self.transform:
            return val

        transform_params = TransformParams(val, selected_element, params.url())
        return self._transform_val(self.transform, transform_params, val)

    @staticmethod
    def _transform_val(transformers, transform_params, val):

        """
        Applies transformations sequentially if multiple transformers exist.
        :param transformers: A list of transformer objects.
        :param transform_params: The parameters passed to the transformers.
        :param val: The value to be transformed.
        :return: The transformed value.
        """
        if isinstance(transformers, list):
            acc = val
            for tr in transformers:
                if tr is not None:
                    acc = tr.transform(transform_params)
                    transform_params.set_val(acc)
            return acc
        elif isinstance(transformers, Transformer):
            return transformers.transform(transform_params)
        return val

File: java/src/main/java/com/purescraps/purehtml/configs/types/PrimitiveValueConfig.java
package com.purescraps.purehtml.configs.types;

import java.util.ArrayList;
import java.util.List;

import com.purescraps.purehtml.backend.PureHTMLNode;
import com.purescraps.purehtml.interfaces.ExtractParams;
import com.purescraps.purehtml.interfaces.GetSelectorMatchesParams;
import com.purescraps.purehtml.transformers.TransformParams;
import com.purescraps.purehtml.transformers.Transformer;

public class PrimitiveValueConfig extends ConfigWithSelector {

    private List<Transformer> transform = new ArrayList<>(); // List to store transformers

    public PrimitiveValueConfig(String selector, List<Transformer> transform) {
        super();
        if (transform != null) {
            this.transform = transform;
        }
        this.selector = selector;
    }

    public Object extract(ExtractParams params) {

        /*
         * Extracts a value from the parameters, applying any transformations if
         * available.
         * 
         * @param params: The ExtractParams object containing context for extraction.
         * 
         * @return: The extracted and transformed value as a String.
         */
        String val = null;

        GetSelectorMatchesParams selectorParams = new GetSelectorMatchesParams() {
            @Override
            public boolean isAlreadyMatched() {
                return params.getElementAlreadyMatched();
            }

            @Override
            public boolean isIncludeRoot() {
                return false;
            }

        };

        PureHTMLNode selectedElement = getFirstMatch(params.node(), selectorParams, params.document());

        if (selectedElement != null) {
            val = selectedElement.text();
        }

        // If no transformation is required, return the value as is
        if (transform == null) {
            return val;
        }

        // Applying transformation
        TransformParams transformParams = new TransformParams(val, selectedElement, params.url());
        return this.transformVal(transform, transformParams, val);
    }

    protected Object transformVal(Object transformer, TransformParams transformParams, Object val) {
        if (transformer instanceof Transformer) {
            return ((Transformer) transformer).transform(transformParams);
        }
        // If transformer is a List of Transformers, apply them sequentially
        if (transformer instanceof List<?>) {
            Object acc = val;
            List<Transformer> transformers = (List<Transformer>) transformer;
            for (Transformer tr : transformers) {
                if (tr != null) {
                    acc = tr.transform(transformParams);
                    transformParams.setVal(acc);
                }
            }
            return acc;
        }
        return transformer;
    }
}
